File: packages/dart/lib/Seguranca/Ataques/AtaqueComplexidadeLinear.dart
import 'dart:math';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Berlekamp-Massey: calcula a complexidade linear (tamanho do menor LFSR que
/// reproduz a sequência de bits). Para uma sequência aleatória de N bits, a
/// complexidade fica perto de N/2. Se a cifra esconder estrutura linear (tipo
/// um LFSR disfarçado), a complexidade cai MUITO abaixo disso - e aí a cifra
/// seria atacável resolvendo um sistema linear em vez de força bruta.
class AtaqueComplexidadeLinear extends Ataque {
  final int _bitsPorAmostra;
  final int _amostras;
  final double _limiteRelativo;

  AtaqueComplexidadeLinear({
    int bitsPorAmostra = 1024,
    int amostras = 5,
    double limiteRelativo = 0.4,
  }) : _bitsPorAmostra = bitsPorAmostra,
       _amostras = amostras,
       _limiteRelativo = limiteRelativo;

  @override
  String nome() => 'Complexidade linear (Berlekamp-Massey)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = alvo.chaveDeTeste();

    final relativos = <double>[];
    var pior = 1.0;
    for (var a = 0; a < _amostras; a++) {
      final bytes = alvo.gerarKeystreamBruto(
        chave,
        randomBytes(alvo.tamanhoIv()),
        'enc',
        _bitsPorAmostra ~/ 8,
      );
      final bits = _paraBits(bytes, _bitsPorAmostra);
      final relativo = _berlekampMassey(bits) / _bitsPorAmostra;
      relativos.add(relativo);
      pior = min(pior, relativo);
    }

    final media = relativos.reduce((a, b) => a + b) / relativos.length;
    final vulneravel = pior < _limiteRelativo;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.critica : Severidade.info,
      'complexidade linear relativa: '
          'média=${(media * 100).toStringAsFixed(1)}%, '
          'pior=${(pior * 100).toStringAsFixed(1)}% de $_bitsPorAmostra bits '
          '(esperado ~50%; limite>=${(_limiteRelativo * 100).toStringAsFixed(0)}%)',
      {'media': media, 'pior': pior},
    );
  }

  List<int> _paraBits(List<int> bytes, int limite) {
    final bits = <int>[];
    for (var i = 0; i < bytes.length && bits.length < limite; i++) {
      final byte = bytes[i];
      for (var b = 7; b >= 0; b--) {
        bits.add((byte >> b) & 1);
      }
    }
    return bits.sublist(0, min(limite, bits.length));
  }

  /// Algoritmo de Berlekamp-Massey sobre GF(2).
  int _berlekampMassey(List<int> s) {
    final n = s.length;
    final c = List<int>.filled(n, 0);
    c[0] = 1;
    final b = List<int>.filled(n, 0);
    b[0] = 1;
    var l = 0;
    var m = -1;

    for (var i = 0; i < n; i++) {
      var d = s[i];
      for (var j = 1; j <= l; j++) {
        d ^= c[j] & s[i - j];
      }

      if (d == 1) {
        final t = List<int>.from(c);
        final shift = i - m;
        for (var j = 0; j + shift < n; j++) {
          c[j + shift] ^= b[j];
        }
        if (l <= i ~/ 2) {
          l = i + 1 - l;
          m = i;
          b.setAll(0, t);
        }
      }
    }

    return l;
  }
}

File: packages/rust/src/seguranca/ataques/ataque_complexidade_linear.rs
//! Berlekamp-Massey: calcula a complexidade linear (tamanho do menor LFSR que
//! reproduz a sequência de bits). Para uma sequência aleatória de N bits, a
//! complexidade fica perto de N/2. Se a cifra esconder estrutura linear (tipo
//! um LFSR disfarçado), a complexidade cai MUITO abaixo disso - e aí a cifra
//! seria atacável resolvendo um sistema linear em vez de força bruta.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueComplexidadeLinear {
    pub bits_por_amostra: usize,
    pub amostras: usize,
    pub limite_relativo: f64,
}

impl AtaqueComplexidadeLinear {
    pub fn new() -> Self {
        AtaqueComplexidadeLinear {
            bits_por_amostra: 1024,
            amostras: 5,
            limite_relativo: 0.4,
        }
    }
}

impl Default for AtaqueComplexidadeLinear {
    fn default() -> Self {
        Self::new()
    }
}

impl AtaqueComplexidadeLinear {
    fn para_bits(bytes: &[u8], limite: usize) -> Vec<u8> {
        let mut bits: Vec<u8> = Vec::new();
        let mut i = 0;
        while i < bytes.len() && bits.len() < limite {
            let byte = bytes[i];
            for b in (0..8).rev() {
                bits.push((byte >> b) & 1);
            }
            i += 1;
        }
        bits.truncate(limite);
        bits
    }

    /// Algoritmo de Berlekamp-Massey sobre GF(2).
    fn berlekamp_massey(s: &[u8]) -> usize {
        let n = s.len();
        let mut c = vec![0u8; n];
        c[0] = 1;
        let mut b = vec![0u8; n];
        b[0] = 1;
        let mut l = 0usize;
        let mut m: i64 = -1;

        for i in 0..n {
            let mut d = s[i];
            for j in 1..=l {
                d ^= c[j] & s[i - j];
            }

            if d == 1 {
                let t = c.clone();
                let shift = (i as i64 - m) as usize;
                for j in 0..n.saturating_sub(shift) {
                    c[j + shift] ^= b[j];
                }
                if l <= i / 2 {
                    l = i + 1 - l;
                    m = i as i64;
                    b = t;
                }
            }
        }

        l
    }
}

impl Ataque for AtaqueComplexidadeLinear {
    fn nome(&self) -> String {
        "Complexidade linear (Berlekamp-Massey)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();

        let mut relativos: Vec<f64> = Vec::with_capacity(self.amostras);
        let mut pior = 1.0f64;
        for _ in 0..self.amostras {
            let bytes = alvo.gerar_keystream_bruto(
                chave.as_bytes(),
                &random_bytes(alvo.tamanho_iv()),
                "enc",
                self.bits_por_amostra / 8,
            );
            let bits = Self::para_bits(&bytes, self.bits_por_amostra);
            let relativo = Self::berlekamp_massey(&bits) as f64 / self.bits_por_amostra as f64;
            relativos.push(relativo);
            pior = pior.min(relativo);
        }

        let media = relativos.iter().sum::<f64>() / relativos.len() as f64;
        let vulneravel = pior < self.limite_relativo;

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Critica
            } else {
                Severidade::Info
            },
            &format!(
                "complexidade linear relativa: média={:.1}%, pior={:.1}% de {} bits (esperado ~50%; limite>={:.0}%)",
                media * 100.0,
                pior * 100.0,
                self.bits_por_amostra,
                self.limite_relativo * 100.0
            ),
        ))
    }
}
